### PR TITLE
Barracks Minimum Stock + Crafting Tweaks

### PR DESCRIPTION
**Pull request by evanpickett**
# Changes proposed in this pull request:
- Minimum stock on barracks (will help late-stage colonies maintain gold for spies)
- Glassblowers can use glass bottles, but not for honey. In vanilla this doesn't do anything, but this pairs well with other mods
- Mechanics can craft diamond blocks back into diamonds

Review please

### CODE REVIEW

**Comment by Raycoms** [comment]
Tested, does it work nicely?

**Comment by evanpickett** [comment]
After some additional testing it seems that some schematics where barracks use racks will have led to situations where couriers would deliver to racks, and barracks didn't check for that (they only looked for gold in their hut blocks). Just committed some changes that should make it where spies can be used even if the gold is in racks rather than the barracks hut block.

File: src/main/resources/data/minecolonies/tags/items/mechanic_product.json [hunk 1]
```diff
@@ -1,31 +1,32 @@
...
+    "minecraft:heavy_weighted_pressure_plate",
+    "minecraft:sticky_piston",
+    "minecraft:trapped_chest",
+    "minecraft:ender_chest",
+    "minecraft:lever",
+    "minecraft:daylight_detector",
+    "minecraft:dried_kelp_block",
+    "minecraft:honey_block",
+    "minecraft:honeycomb_block",
+    "minecraft:diamond"
```
Comment (ravenbuilder934) [inline src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: Why did you add diamonds here? I don't think other storage blocks can be uncrafted by the mechanic

Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: Do you think I should move that over to the blacksmith then?

Comment (ravenbuilder934) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: Yea blacksmith, to match with the rest

File: src/main/resources/data/minecolonies/tags/items/mechanic_product.json [hunk 2]
```diff
@@ -1,31 +1,31 @@
...
+    "minecraft:light_weighted_pressure_plate",
+    "minecraft:heavy_weighted_pressure_plate",
+    "minecraft:sticky_piston",
+    "minecraft:trapped_chest",
+    "minecraft:ender_chest",
+    "minecraft:lever",
+    "minecraft:daylight_detector",
+    "minecraft:dried_kelp_block",
+    "minecraft:honey_block",
+    "minecraft:honeycomb_block"
```
Comment (ravenbuilder934) [inline src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: This file still has changes?

File: src/main/resources/data/minecolonies/tags/items/blacksmith_product.json [hunk 3]
```diff
@@ -1,8 +1,9 @@
...
-    "minecraft:shears"
-  ]
-}
+{
+  "replace": false,
+  "values": [
+    "#forge:nuggets",
+    "#forge:ingots",
+    "minecraft:shears",
+    "minecraft:diamond"
```
Comment (ravenbuilder934) [inline src/main/resources/data/minecolonies/tags/items/blacksmith_product.json]: Maybe add the storage blocks tag to the blacksmith_ingredient instead?

File: src/main/resources/data/minecolonies/tags/items/mechanic_product.json (hunk 2 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: not sure what's going on with that, looks like that's the case with blacksmith_products too despite there being no actual changes. Maybe it's because I switched from my laptop to desktop halfway through?

File: src/main/resources/data/minecolonies/tags/items/blacksmith_product.json (hunk 3 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/blacksmith_product.json]: good idea, made that change

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java [hunk 4]
```diff
@@ -157,10 +166,10 @@ public void onColonyTick(@NotNull final IColony colony)
         {
             if (!colony.getRaiderManager().areSpiesEnabled())
             {
-                final int amount = InventoryUtils.getItemCountInItemHandler(this.getTileEntity().getInventory(), Items.GOLD_INGOT);
+                final int amount = InventoryUtils.getItemCountInItemHandlers(getHandlers(), i -> i.getItem().equals(Items.GOLD_INGOT));
                 if (amount >= SPIES_GOLD_COST)
                 {
-                    InventoryUtils.removeStackFromItemHandler(tileEntity.getInventory(), new ItemStack(Items.GOLD_INGOT, SPIES_GOLD_COST), SPIES_GOLD_COST);
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: If the barracks has a list of containers, then tileEntity.getInventory should be the combined itemHandler of all them and not cause problems. If this is not the case, we maybe need the barracks tileEntity to follow the same handler code as the others. But I beleive that should not be the case.

File: src/main/resources/data/minecolonies/tags/items/mechanic_product.json (hunk 2 above)
Comment (ravenbuilder934) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: could be line endings, could be tabs/spaces auto-reformatting, could be something else. if you're using gitkraken I know a way to revert the changes

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java (hunk 4 above)
Comment (evanpickett) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: I felt that should be the case but ultimately when I use tileEntity.getInventory() it only checks the hut block's inventory for the gold rather than the racks. Maybe this isn't the most graceful solution but I'm not sure exactly where that problem lies otherwise. Any ideas?

File: src/main/resources/data/minecolonies/tags/items/mechanic_product.json (hunk 2 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: Just started using gitkraken recently, if you could let me know that would be great! Maybe message me on Discord if you'd like to avoid cluttering up the comments? `Brick_man#0096`

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java (hunk 4 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: ohhh, yeah it shouldn't be getInventory, it should be getCapability with the itemhandler

Comment (evanpickett) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: That seems to have done the trick!

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: We can even make this more simple tryRemoveStackFromItemHandler(itemhandler, new ItemStack(gold, count)) 
if that returns true, set enabled.

File: src/api/java/com/minecolonies/api/util/InventoryUtils.java [hunk 5]
```diff
@@ -2217,6 +2217,37 @@ public static void removeStackFromItemHandler(final IItemHandler handler, final
...
+     */
+    public static void removeStackFromItemHandlers(final List<IItemHandler> handlers, final ItemStack input, final int count)
+    {
+        final ItemStack workingStack = input.copy();
+        int localCount = count;
+        int tries = 0;
+        for (final IItemHandler handler : handlers) {
+            while (tries < count) {
+                final int slot = findFirstSlotInItemHandlerNotEmptyWith(handler, stack -> ItemStackUtils.compareItemStacksIgnoreStackSize(workingStack, stack));
+                if (slot == -1) {
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/util/InventoryUtils.java]: this can go now

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java (hunk 4 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: Also, please setup our formatter for this

Comment (evanpickett) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java:163]: I'm now using the tryRemoveStackFromItemHandler, seems much more compact

File: src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json [hunk 6]
```diff
@@ -1,7 +1,8 @@
...
-  ]
-}
\ No newline at end of file
+{
+  "replace": false,
+  "values": [
+    "#forge:nuggets",
+    "#forge:ingots",
+    "#forge:storage_blocks"
+  ]
```
Comment (Raycoms) [inline src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json:8]: Hmm, this allow to uncraft ALL storage blocks, Either we need sth like "metal_blocks" or we need sth else.

File: src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json [hunk 7]
```diff
@@ -1,7 +1,8 @@
...
-    "#forge:glass_panes"
-  ]
-}
+{
+  "replace": false,
+  "values": [
+    "#forge:glass",
+    "#forge:glass_panes",
+    "minecraft:glass_bottle"
+  ]
```
Comment (Raycoms) [inline src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json]: What is the reason for this?

File: src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json (hunk 6 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json:8]: Not sure if a tag like that exists, and that would defeat the purpose of that change originally which was to allow someone to change diamond blocks back to diamonds (because I'm not sure you could classify diamonds as metal). Additionally, I think it would be helpful if someone could turn emerald blocks back to emeralds. So if you're opposed to all storage blocks, then perhaps I should just put in diamonds and emeralds directly into the blacksmith_products and leave it at that?

File: src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json (hunk 7 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json]: Figured it was a harmless way to add some mod compatibility for mods which use glass bottles as recipes. If you feel this is outside the scope of what should be included in this PR I can remove it. I'm more concerned about the gem issue (diamonds specifically) anyways

File: src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json (hunk 6 above)
Comment (Raycoms) [reply src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json:8]: Yeah, I'm fine with Diamond_block and emerald_block

File: src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json (hunk 7 above)
Comment (Raycoms) [reply src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json]: I feel it could allow for all kind of recipes like filling up bottles with sth.

Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json]: In theory yes, it could lead to undesirable recipes for the glassblower. Maybe it should be handled on a case-by-case basis with other mods for compatibility. I can scrap this change if that's more aligned with how you feel crafting should work.

**Review by someaddons — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java [hunk 8]
```diff
@@ -27,14 +29,20 @@
 import net.minecraft.util.math.BlockPos;
 import net.minecraft.world.World;
 import net.minecraftforge.common.util.Constants;
+import net.minecraftforge.common.util.LazyOptional;
+import net.minecraftforge.items.CapabilityItemHandler;
+import net.minecraftforge.items.IItemHandler;
 import org.jetbrains.annotations.NotNull;
 
 import java.util.ArrayList;
+import java.util.HashSet;
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java]: I think there are some imports to clean up left here

File: src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json (hunk 6 above)
Comment (Raycoms) [reply src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json:8]: Waiting for those yet

File: src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json (hunk 7 above)
Comment (Raycoms) [reply src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json]: yeah, better scrap this

File: src/main/resources/data/minecolonies/tags/items/mechanic_product.json (hunk 2 above)
Comment (ravenbuilder934) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: If you right-click on the commit there's an option to "revert" it, you can choose to only revert certain files/etc

Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/mechanic_product.json]: thanks!

File: src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json (hunk 6 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/blacksmith_ingredient.json:8]: added diamond_block and emerald_block into the blacksmith_ingredient

File: src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json (hunk 7 above)
Comment (evanpickett) [reply src/main/resources/data/minecolonies/tags/items/glassblower_ingredient.json]: done, reverted

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java (hunk 8 above)
Comment (evanpickett) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingBarracks.java]: shoulllld be good now

**Review by Raycoms — APPROVED** [review]

**Review by ravenbuilder934 — APPROVED** [review]
